### ISSUE TITLE
Produce() dropping messages?

### ISSUE THREAD

**Issue by jon-droniak** [confluentinc/confluent-kafka-dotnet#1293]
Description
===========
Hello,

My team currently has been experiencing an issue where messages seem to be getting lost or otherwise not delivered. We use Confluent for both producing and consuming messages, and delivery guarantees are a must for our use case. We have been investigating this issue for quite some time and we cannot seem to figure out why we keep losing messages. The interesting part is that there is no error being thrown that we can see in our logs or the broker logs. We know this is happening because we see evidence in our logs that show a message was sent to the topic, but there was never a consume associated with that message. Our producer currently uses Produce (not ProduceAsync).

We have also gone through our broker logs. We don't see any evidence that the message actually reached the brokers, yet our service that produces messages is logging a successful response.

We do have logging in the delivery handler for Produce when there is an error, but we have not once seen that log. I suspect that this may be due to #1094.

Our apps use Confluent version 1.4.2 and our Kafka brokers are running version 2.5. The apps we have that use Confluent are hosted on Windows Server 2016 VMs.

Our producer configuration is as follows:

- EnableIdempotence: true
- Acks: All
- Partitioner: ConsistentRandom

Other configuration values are left at default.

In addition we have also tried setting BatchNumMessages and QueueBufferingMaxMessages both to 1 to rule out any issues with the in-memory queue, however we were still losing messages with this configuration.

How to reproduce
================
Unfortunately this issue cannot be easily reproduced since it is sporadic in nature and does not happen consistently. We can go several days to sometimes a week without losing a message.

Checklist
=========

Please provide the following information:
 - [ ] A complete (i.e. we can run it), minimal program demonstrating the problem. No need to supply a project file.
 - [ ] Confluent.Kafka nuget version.
 - [ ] Apache Kafka version.
 - [ ] Client configuration.
 - [ ] Operating system.
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration).
 - [ ] Provide broker log excerpts.
 - [ ] Critical issue.

**Comment by mhowlett** [comment]
i think it's extremely unlikely that if you have acks set to all, you would get a successful delivery result in the case a message was not successfully persisted to the cluster. i think this functionality is very unlikely to be broken - i would definitely be targeting your attention elsewhere in looking for the root cause of the problem.

> We don't see any evidence that the message actually reached the brokers

how are you determining that? 

#1094 has been addressed in v1.4.0 - exceptions thrown in produce handlers will be surfaced as error events on the error handler set via the `SetErrorHandler` builder class method.

**Comment by jon-droniak** [comment]
Thanks for your response.

We have updated our error handling using `SetErrorHandler` and for reference I've added it below:
```
if (error.Code == ErrorCode.NoError)
{
    return;
}

_logger.LogWarning($"KafkaPublisher: An exception occurred while sending to Kafka brokerUrls={brokerUrls}, topic={configuration.Topic}, messageContent={messageContent}",
                        error.Reason);
```

We did encounter this issue a few more times yesterday with the error handling updates in our environment, and we did not see any publish errors in our logs.

> how are you determining that?

We have used several methods to investigate this, listed below:

1. Read the raw Kafka topics using a console consumer
2. Used an on-demand automated test that reads from a topic with a unique group ID
3. Examined our logs and saw that our messaging service logged a request to produce a message to the topic and found that there is no consume associated with that message. Basically our logs show that there was a successful produce, but no consume.

**Comment by jon-droniak** [comment]
I have a quick update for this ticket:

We've found the root cause of this issue. It turns out that we were creating producers on every request to our service, making it effectively a transient producer. We've updated our producer to be created once and only once during the startup of our service, and this has resolved the issue.

Apologies for wasting time, we thought we had tried everything we could.
